### PR TITLE
Add generate_yaml command to easily test KubernetesExecutor before deploying pods

### PR DESCRIPTION
**Pull request by dimberman**
To aide in migration testing (and general testing going forward)
we're adding a dry-run command to the airflow CLI. This command will
allow users to see the YAML of a KubernetesExecutor task without running it
in Kubernetes.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/cli/cli_parser.py [hunk 1]
```diff
@@ -142,6 +142,10 @@ def positive_int(value):
     ("-e", "--end-date"),
     help="Override end_date YYYY-MM-DD",
     type=parsedate)
+ARG_OUTPUT_PATH = Arg(
+    ("-o", "--output-path",),
+    help="The output for generated yaml files",
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py:158]:
```suggestion
    help="The output for generated yaml files",
    type=argparse.FileType('w')
```
See: https://docs.python.org/3/library/argparse.html#filetype-objects

File: airflow/cli/cli_parser.py [hunk 2]
```diff
@@ -778,6 +782,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.dag_command.dag_list_dags'),
         args=(ARG_SUBDIR, ARG_OUTPUT),
     ),
+    ActionCommand(
+        name='generate_yaml',
+        help="Generate YAML files for all tasks in DAG",
+        func=lazy_load_command('airflow.cli.commands.dag_command.generate_pod_yaml'),
+        args=(ARG_SUBDIR, ARG_DAG_ID_OPT,ARG_OUTPUT_PATH),
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]: Is the DAG ID really optional?

File: airflow/cli/commands/dag_command.py [hunk 3]
```diff
@@ -378,6 +379,47 @@ def dag_list_dag_runs(args, dag=None):
...
+        import yaml
+        from kubernetes.client.api_client import ApiClient
+        api_client = ApiClient()
+        date_string = pod_generator.datetime_to_label_safe_datestring(execution_date)
+        yaml_file_name = f"{args.dag_id}_{ti.task_id}_{date_string}.yml"
+        os.makedirs(os.path.dirname(yaml_output_path), exist_ok=True)
+        with open(yaml_output_path + yaml_file_name, "w") as output:
+            sanitized_pod = api_client.sanitize_for_serialization(pod)
+            output.write(yaml.dump(sanitized_pod))
+    print(f"YAML output can be found at {yaml_output_path}")
```
Comment (mik-laj) [inline airflow/cli/commands/dag_command.py]:
```suggestion
    print(f"YAML output can be found at {yaml_output_path}", file=sys.err)
```

File: airflow/cli/cli_parser.py (hunk 2 above)
Comment (kaxil) [reply airflow/cli/cli_parser.py]: might be an error, should not be optional

File: airflow/cli/commands/dag_command.py [hunk 4]
```diff
@@ -378,6 +379,47 @@ def dag_list_dag_runs(args, dag=None):
...
+def generate_pod_yaml(args, session=None):
+    """Generates yaml files for each task in the DAG. Used for testing output of KubernetesExecutor"""
+
+    from airflow.executors.kubernetes_executor import AirflowKubernetesScheduler, KubeConfig
+    from airflow.kubernetes import pod_generator
+    from airflow.kubernetes.pod_generator import PodGenerator
+    from airflow.kubernetes.worker_configuration import WorkerConfiguration
+    execution_date = datetime.datetime(2020,11,3)
+    dag = get_dag(subdir=args.subdir, dag_id=args.dag_id)
+    yaml_output_path = args.output_path or "/tmp/airflow_generated_yaml/"
```
Comment (mik-laj) [inline airflow/cli/commands/dag_command.py]: Should be `sys.stdout` by default?

File: tests/cli/commands/test_dag_command.py [hunk 5]
```diff
@@ -139,6 +139,12 @@ def test_show_dag_print(self):
         self.assertIn("graph [label=example_bash_operator labelloc=t rankdir=LR]", out)
         self.assertIn("runme_2 -> run_after_loop", out)
 
+    def test_generate_dag_yaml(self):
+        with contextlib.redirect_stdout(io.StringIO()) as temp_stdout:
+            dag_command.generate_pod_yaml(self.parser.parse_args([
+                'dags', 'generate_yaml', '--dag-id', 'example_bash_operator']))
+
```
Comment (mik-laj) [inline tests/cli/commands/test_dag_command.py]: Can you add even one assertion?

File: airflow/cli/commands/dag_command.py (hunk 4 above)
Comment (dimberman) [reply airflow/cli/commands/dag_command.py]: @mik-laj would people want that as stdout? The idea is that it would create a file per-task.

File: tests/cli/commands/test_dag_command.py (hunk 5 above)
Comment (dimberman) [reply tests/cli/commands/test_dag_command.py]: @mik-laj yes I plan to add assertions that a yaml file is created and that it's a valid pod.

File: airflow/cli/cli_parser.py (hunk 2 above)
Comment (dimberman) [reply airflow/cli/cli_parser.py]: How do I make this not optional?

File: airflow/cli/cli_parser.py [hunk 6]
```diff
@@ -778,6 +783,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.dag_command.dag_list_dags'),
         args=(ARG_SUBDIR, ARG_OUTPUT),
     ),
+    ActionCommand(
+        name='generate_yaml',
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]: What do you think to add this to the new ``kubernetes`` group? We currently have sections for celery executor.

```
Celery components

  airflow celery flower - Start a Celery Flower
  airflow celery stop - Stop the Celery worker gracefully
  airflow celery worker - Start a Celery worker node
```

File: airflow/cli/cli_parser.py [hunk 7]
```diff
@@ -778,6 +783,12 @@ class GroupCommand(NamedTuple):
...
+        help="Generate YAML files for all tasks in DAG",
+        func=lazy_load_command('airflow.cli.commands.dag_command.generate_pod_yaml'),
+        args=(ARG_SUBDIR, ARG_DAG_ID_OPT,ARG_OUTPUT_PATH),
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]:
```suggestion
        args=(ARG_SUBDIR, ARG_DAG_ID, ARG_OUTPUT_PATH),
```

File: airflow/cli/commands/dag_command.py (hunk 4 above)
Comment (mik-laj) [reply airflow/cli/commands/dag_command.py]: Why one task in a separate file? Wouldn't it be more convenient if there was one file for each DAG? This will prevent you from having to write the data to a file, but you can display it on the screen.  Ideally, the YAML displayed on the screen should also have colors. Take a look: `airflow config list`.

Comment (dimberman) [reply airflow/cli/commands/dag_command.py]: @mik-laj because usually if I'm debugging I'm debugging a specific task. We can add a "join_in_one_file" option if there is interest, but that can be a separate commit.

File: tests/cli/commands/test_dag_command.py (hunk 5 above)
Comment (dimberman) [reply tests/cli/commands/test_dag_command.py]: @mik-laj PTAL added assertions

File: airflow/cli/cli_parser.py (hunk 6 above)
Comment (dimberman) [reply airflow/cli/cli_parser.py]: @mik-laj so create a section for one command? Seems like a bit of overkill...

File: airflow/cli/cli_parser.py (hunk 1 above)
Comment (kaxil) [reply airflow/cli/cli_parser.py:158]: This hasn't been fixed @dimberman

File: airflow/cli/commands/dag_command.py [hunk 8]
```diff
@@ -378,6 +379,47 @@ def dag_list_dag_runs(args, dag=None):
...
 
+@cli_utils.action_logging
+def generate_pod_yaml(args):
+    """Generates yaml files for each task in the DAG. Used for testing output of KubernetesExecutor"""
+
+    from airflow.executors.kubernetes_executor import AirflowKubernetesScheduler, KubeConfig
+    from airflow.kubernetes import pod_generator
+    from airflow.kubernetes.pod_generator import PodGenerator
+    from airflow.kubernetes.worker_configuration import WorkerConfiguration
+    execution_date = datetime.datetime(2020, 11, 3)
```
Comment (kaxil) [inline airflow/cli/commands/dag_command.py]: Why is the execution_date hardcoded ?

File: airflow/cli/commands/dag_command.py [hunk 9]
```diff
@@ -378,6 +379,47 @@ def dag_list_dag_runs(args, dag=None):
...
+            command=ti.command_as_list(),
+            kube_executor_config=PodGenerator.from_obj(ti.executor_config),
+            worker_uuid="worker-config",
+            namespace=kube_config.executor_namespace,
+            worker_config=WorkerConfiguration(kube_config=kube_config).as_pod()
+        )
+        import os
+
+        import yaml
+        from kubernetes.client.api_client import ApiClient
```
Comment (kaxil) [inline airflow/cli/commands/dag_command.py]: Why are these here ??? let's properly format them (os and yaml can go at the top of the file)

File: tests/cli/commands/test_dag_command.py [hunk 10]
```diff
@@ -139,6 +140,19 @@ def test_show_dag_print(self):
...
+        directory = "/tmp/airflow_dry_run_test/"
+        file_name = "example_bash_operator_run_after_loop_2020-11-03T00_00_00.yml"
+        if os.path.exists(directory):
+            shutil.rmtree(directory)
+        dag_command.generate_pod_yaml(self.parser.parse_args([
+            'dags', 'generate_kubernetes_pod_yaml',
+            "--output-path", directory, 'example_bash_operator']))
+        self.assertEqual(len(os.listdir(directory)), 6)
+        self.assertTrue(os.path.isfile(directory + file_name))
+        self.assertNotEqual(os.stat(directory + file_name).st_size, 0)
```
Comment (kaxil) [inline tests/cli/commands/test_dag_command.py]: you can use `assertGreater` here https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertGreater

File: airflow/cli/commands/dag_command.py (hunk 8 above)
Comment (dimberman) [reply airflow/cli/commands/dag_command.py]: @kaxil since this endpoint is for testing purposes, it didn't make sense to force users to put in an execution date, and it seemed to make sense to hard code it so there would be consistency. Do you feel differently?

Comment (kaxil) [reply airflow/cli/commands/dag_command.py]: I would still think we can use `timezone.utcnow()` as execution date instead of `datetime.datetime(2020, 11, 3)`

where timezone.utcnow is https://github.com/apache/airflow/blob/master/airflow/utils/timezone.py#L52

File: airflow/cli/commands/dag_command.py [hunk 11]
```diff
@@ -378,6 +380,48 @@ def dag_list_dag_runs(args, dag=None):
...
+@cli_utils.action_logging
+def generate_pod_yaml(args):
+    """Generates yaml files for each task in the DAG. Used for testing output of KubernetesExecutor"""
+
+    from airflow.executors.kubernetes_executor import AirflowKubernetesScheduler, KubeConfig
+    from airflow.kubernetes import pod_generator
+    from airflow.kubernetes.pod_generator import PodGenerator
+    from airflow.kubernetes.worker_configuration import WorkerConfiguration
+    from airflow.settings import pod_mutation_hook
+    from airflow.utils import timezone
```
Comment (kaxil) [inline airflow/cli/commands/dag_command.py]: This imports can go at the top of the file, no ?

File: airflow/cli/commands/dag_command.py [hunk 12]
```diff
@@ -378,6 +380,48 @@ def dag_list_dag_runs(args, dag=None):
...
+            date=ti.execution_date,
+            command=ti.command_as_list(),
+            kube_executor_config=PodGenerator.from_obj(ti.executor_config),
+            worker_uuid="worker-config",
+            namespace=kube_config.executor_namespace,
+            worker_config=WorkerConfiguration(kube_config=kube_config).as_pod()
+        )
+        pod_mutation_hook(pod)
+
+        from kubernetes.client.api_client import ApiClient
```
Comment (kaxil) [inline airflow/cli/commands/dag_command.py]: same here, import in the middle of the file / function ? Any particular reason?

Comment (dimberman) [reply airflow/cli/commands/dag_command.py]: @kaxil this one has a reason: to avoid causing errors if a user does not have the kubernetes client installed.

Comment (kaxil) [reply airflow/cli/commands/dag_command.py]: It can be at Line 392, it will still be inside the function

**Review by kaxil — APPROVED** [review]

File: tests/cli/commands/test_dag_command.py [hunk 13]
```diff
@@ -139,6 +140,19 @@ def test_show_dag_print(self):
         self.assertIn("graph [label=example_bash_operator labelloc=t rankdir=LR]", out)
         self.assertIn("runme_2 -> run_after_loop", out)
 
+    def test_generate_dag_yaml(self):
+
+        directory = "/tmp/airflow_dry_run_test/"
```
Comment (mik-laj) [inline tests/cli/commands/test_dag_command.py]: Can you use [TemporaryDirectory](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory) to avoid side-effect?

File: airflow/cli/commands/dag_command.py [hunk 14]
```diff
@@ -378,6 +380,48 @@ def dag_list_dag_runs(args, dag=None):
     print(table)
 
 
+@cli_utils.action_logging
+def generate_pod_yaml(args):
+    """Generates yaml files for each task in the DAG. Used for testing output of KubernetesExecutor"""
+
+    from kubernetes.client.api_client import ApiClient
```
Comment (mik-laj) [inline airflow/cli/commands/dag_command.py:387]: This is a command that is intended for the end user. Do you think it's worth adding a friendly message that will be displayed when Kubernetes is not installed or not configured?

File: airflow/cli/cli_parser.py (hunk 6 above)
Comment (mik-laj) [reply airflow/cli/cli_parser.py]: Ok. In the future, we can move this to a separate group.

File: airflow/cli/commands/dag_command.py (hunk 8 above)
Comment (mik-laj) [reply airflow/cli/commands/dag_command.py]: Elsewhere execution_date is required. Should we unify it?
```
usage: airflow tasks test [-h] [-n] [--env-vars ENV_VARS] [-m] [-S SUBDIR]
                          [-t TASK_PARAMS]
                          dag_id task_id execution_date
```
```
airflow dags test [-h] [--imgcat-dagrun] [--save-dagrun SAVE_DAGRUN]
                         [--show-dagrun] [-S SUBDIR]
                         dag_id execution_date
```
```
usage: airflow tasks run [-h] [--cfg-path CFG_PATH] [-f] [-A] [-i] [-I] [-N]
                         [-l] [-m] [-p PICKLE] [--pool POOL] [--ship-dag]
                         [-S SUBDIR]
                         dag_id task_id execution_date
```

File: airflow/cli/cli_parser.py [hunk 15]
```diff
@@ -142,6 +142,10 @@ def positive_int(value):
     ("-e", "--end-date"),
     help="Override end_date YYYY-MM-DD",
     type=parsedate)
+ARG_OUTPUT_PATH = Arg(
+    ("-o", "--output-path",),
+    help="The output for generated yaml files",
+    default="/tmp/airflow_yaml_output/")
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]: Wouldn't it be better to write in Current Working Directory? This will provide easier access to these files. Most linux utilities use CWD by default.
```
tar xf pathtoarchive/archive.tar

```

File: airflow/cli/commands/dag_command.py (hunk 14 above)
Comment (mik-laj) [reply airflow/cli/commands/dag_command.py:387]: We have a similar check for Celery.
Screenshot 2020-09-03 at 23 10 08

File: airflow/cli/cli_parser.py [hunk 16]
```diff
@@ -778,6 +782,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.dag_command.dag_list_dags'),
         args=(ARG_SUBDIR, ARG_OUTPUT),
     ),
+    ActionCommand(
+        name='generate_kubernetes_pod_yaml',
+        help="Generate YAML files for all tasks in DAG",
+        func=lazy_load_command('airflow.cli.commands.dag_command.generate_pod_yaml'),
+        args=(ARG_SUBDIR, ARG_DAG_ID, ARG_OUTPUT_PATH, ARG_EXEC_DATE),
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]:
```suggestion
        args=(ARG_SUBDIR, ARG_DAG_ID, ARG_OUTPUT_PATH, ARG_EXECUTION_DATE),
```
This will surely make this argument required.

Comment (dimberman) [reply airflow/cli/cli_parser.py]: @mik-laj  does it need to be required? this is just a diagnostic tool so it shouldn't make a difference if we use a dummy execution_date.

File: airflow/cli/cli_parser.py (hunk 15 above)
Comment (dimberman) [reply airflow/cli/cli_parser.py]: By putting it in /tmp we can assure that it will eventually be deleted as the system needs space.

File: airflow/cli/commands/dag_command.py (hunk 14 above)
Comment (dimberman) [reply airflow/cli/commands/dag_command.py:387]: @mik-laj good idea will implement.

File: airflow/cli/cli_parser.py (hunk 15 above)
Comment (mik-laj) [reply airflow/cli/cli_parser.py]: I understand the reasons, but I'm afraid it might be more difficult to use because of this.  It is common to save files in the current directory by default.  If the user wishes to write to tmp, they can execute the following sequence of commands.
```bash
cd /tmp/
mkdir -p airflow_tmp
airflow dags generate_kubernetes_pod_yaml example_bash_operator
ls
```
If we save in /tmp/ the user may feel confused. When I unpack or create a file, I expect results to be visible after executing ``ls``
```bash
cd ~
airflow dags generate_kubernetes_pod_yaml example_bash_operator
ls
# Generated files not visible.
```
In 2020, users have large enough hard drives that we do not have to worry about saving too much data. I think their convenience and productivity are more important.

File: airflow/cli/cli_parser.py (hunk 16 above)
Comment (mik-laj) [reply airflow/cli/cli_parser.py]: This is not necessary, but should we update the other commands as well? Using CLI is very often based on habits, and I think it's worth making sure that all commands have a similar style.
```
airflow dags test example_bash_operator 2019-01-01
airflow dags generate_kubernetes_pod_yaml example_bash_operator --exec-date 2019-01-01
```
I suspect these two commands may be used by the same people and follow two different conventions. This is very confusing to the user. I am in favor of making this argument optional, but it should be done in all commands, or none of them.

Comment (dimberman) [reply airflow/cli/cli_parser.py]: IMHO execution_date should never be a positional argument. However making this change would be way outside of the scope of this ticket. Can we merge this as is and open a thread on the mailing list for a wider discussion of this?

Comment (mik-laj) [reply airflow/cli/cli_parser.py]: Is there any reason why the command `airflow dags generate_kubernetes_pod_yaml` should have exeuction_date optional, and the command `airflow dags test` has this argument required? In my opinion, these two commands have a similar role. These are commands created for development.

File: airflow/cli/cli_parser.py [hunk 17]
```diff
@@ -778,6 +782,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.dag_command.dag_list_dags'),
         args=(ARG_SUBDIR, ARG_OUTPUT),
     ),
+    ActionCommand(
+        name='generate_kubernetes_pod_yaml',
+        help="Generate YAML files for all tasks in DAG",
+        func=lazy_load_command('airflow.cli.commands.dag_command.generate_pod_yaml'),
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]:
```suggestion
        name='dump-pod-yaml',
        help="Generate YAML files with K8S Pod definitions.",
        description="Generate YAML files with Kubernetes Pod definitions for all tasks in DAG",
```
This command name is too long. It doesn't look good in CLI.
Screenshot 2020-09-04 at 00 25 43

File: airflow/cli/cli_parser.py [hunk 18]
```diff
@@ -778,6 +782,12 @@ class GroupCommand(NamedTuple):
         func=lazy_load_command('airflow.cli.commands.dag_command.dag_list_dags'),
         args=(ARG_SUBDIR, ARG_OUTPUT),
     ),
+    ActionCommand(
+        name='generate_kubernetes_pod_yaml',
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]:
```suggestion
        name='generate-kubernetes-pod-yaml',
```
Commands should use dash as a separator. See: https://github.com/apache/airflow/pull/10669
I forgot one command yet, but I will correct it in a moment.

Comment (kaxil) [reply airflow/cli/cli_parser.py]: Good point

Comment (kaxil) [reply airflow/cli/cli_parser.py]: Actually looks like "next_execution" might need changing as well. We definitely need to decide dash or underscore

File: airflow/cli/cli_parser.py (hunk 15 above)
Comment (dimberman) [reply airflow/cli/cli_parser.py]: @kaxil WDYT?

Comment (kaxil) [reply airflow/cli/cli_parser.py]: @mik-laj has a Valid point

Comment (dimberman) [reply airflow/cli/cli_parser.py]: I would personally be really annoyed if I ran this command and then suddenly had to clean up 20 yaml files in my current directory

Comment (kaxil) [reply airflow/cli/cli_parser.py]: I agree with Kamil on this.

Comment (mik-laj) [reply airflow/cli/cli_parser.py]: @potiuk WDYT

Comment (kaxil) [reply airflow/cli/cli_parser.py]: You could output to a directory in the current directory. That way you just need to delete the folder if that is the concern

File: airflow/cli/cli_parser.py (hunk 18 above)
Comment (mik-laj) [reply airflow/cli/cli_parser.py]: PR: https://github.com/apache/airflow/pull/10720

File: airflow/cli/cli_parser.py (hunk 15 above)
Comment (kaxil) [reply airflow/cli/cli_parser.py]: i.e the subcommand can create a folder example "output_yaml_files" or something in the current directory that stores all the yaml's

Comment (dimberman) [reply airflow/cli/cli_parser.py]: Done :)

File: airflow/cli/cli_parser.py [hunk 19]
```diff
@@ -66,6 +66,17 @@ def _check_value(self, action, value):
         if value == 'celery' and executor != ExecutorLoader.CELERY_EXECUTOR:
             message = f'celery subcommand works only with CeleryExecutor, your current executor: {executor}'
             raise ArgumentError(action, message)
+        if value == 'kubernetes':
+            try:
+                from kubernetes.client import models
+                if not models:
+                    message = 'kubernetes subcommand requires that ' \
```
Comment (mik-laj) [inline airflow/cli/cli_parser.py]: Can you add full pip install command - ``pip install 'apache-airflow[cncf.kubernetes]' ``?

**Comment by mik-laj** [comment]
Last thing, can you add more docs? I am thinking of a sentence or two that shows an example use case. Without it, this command will only be available in the reference documentation, and it is not very user-friendly.
https://airflow.readthedocs.io/en/latest/executor/kubernetes.html

**Review by mik-laj — APPROVED** [review]
